internal/runner/minio_test.go

```diff
@@ -132,13 +198,219 @@ func s3AnonAccess(t *testing.T, mts *minio_local.MinioTestServer, logger *log.Lo
 	}
 }
 
+type testCase struct {
+	bucket string
+	size   int64
+	keys   []string
+}
+
+func limitTest(ctx context.Context, t *testing.T, mts *minio_local.MinioTestServer, creds request.AWSCredential, test testCase, logger *log.Logger) {
+
+	// Create a new TMPDIR so that we can cleanup
+	tmpDir, errGo := ioutil.TempDir("", "")
+	if errGo != nil {
+		t.Fatal(kv.Wrap(errGo).With("stack", stack.Trace().TrimRuntime()))
+	}
+	defer func() {
+		os.RemoveAll(tmpDir)
+	}()
+
+	env := map[string]string{}
+
+	authS3, err := s3.NewS3storage(ctx, creds, env, mts.Address, test.bucket, "", false, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		authS3.Close()
+	}()
+
+	keys := make(map[string]int64, len(test.keys))
+
+	// First we try downloading using the test case for positive case
+	firstKey := ""
+	maxBytes := int64(test.size)
+	for _, key := range test.keys {
+		size, warns, err := authS3.Fetch(ctx, key, false, tmpDir, maxBytes, nil)
+		if err != nil {
+			t.Fatal(err, "bucket", test.bucket, "key", key)
+		}
+		if len(warns) != 0 {
+			t.Fatal(warns, "bucket", test.bucket, "key", key)
+		}
+		maxBytes -= size
+		keys[key] = size
+		if len(firstKey) == 0 {
+			firstKey = key
+		}
+	}
+
+	// Now examine the blobs that got downloaded and make sure they are all there
+	files, errGo := os.ReadDir(tmpDir)
+	if errGo != nil {
+		t.Fatal(errGo)
+	}
+
+	for _, file := range files {
+		if _, isPresent := keys[file.Name()]; !isPresent {
+			t.Fatal("unexpected key was downloaded", file.Name(), "keys", spew.Sdump(keys))
+		}
+		if _, isPresent := keys[file.Name()]; !isPresent {
+			t.Fatal("unexpected key was downloaded", file.Name(), "dir", tmpDir, "requested", test.size, "keys", spew.Sdump(keys))
+		}
```

Comment:
Why do we do this twice?
"if _, isPresent" thing?

Reply:
A duplication.  When extracting from maps in go you can specify a second variable that tells you if the key is present.